File: sdks/swift/Sources/DittoChatCore/DittoChat.swift
//
//  File.swift
//  DittoChat
//
//  Created by Erik Everson on 12/2/24.
//

import Foundation
import DittoSwift
import Combine
import UIKit

@MainActor
public protocol DittoSwiftChat {
    func createRoom(withConfig: RoomConfig) async throws -> String
    func createMessage(withConfig: MessageConfig) async throws
    func setCurrentUser(withConfig: UserConfig)

    var publicRoomsPublisher: AnyPublisher<[Room], Never> { get }
    func readRoomById(id: String) async throws -> Room
    func allUsersPublisher() -> AnyPublisher<[ChatUser], Never>

    func updateRoom(room: Room) async throws

    func logout()
}

public struct RoomConfig {
    public let id: String?
    public let name: String
    public let isGenerated: Bool

    public init(id: String? = nil, name: String, isGenerated: Bool = false) {
        self.id = id
        self.name = name
        self.isGenerated = isGenerated
    }
}

public struct MessageConfig {
    public let roomId: String
    public let message: String

    public init(roomId: String, message: String) {
        self.roomId = roomId
        self.message = message
    }
}

public struct UserConfig {
    public let id: String

    public init (id: String) {
        self.id = id
    }
}

// TODO: Hook this up to actually work
public struct ChatRetentionPolicy {
    public var days: Int

    public init(days: Int) {
        self.days = days
    }
}

struct AdminRole: DittoDecodable {
    init(value: [String : Any?]) {
        _id = value["_id"] as? String
        email = value["email"] as? String
    }

    public var _id: String?
    public var email: String?
}

@MainActor
public class DittoChat: DittoSwiftChat, ObservableObject {
    @Published private(set) public var publicRoomsPublisher: AnyPublisher<[Room], Never>
    public var retentionPolicy: ChatRetentionPolicy = .init(days: 30)
    public var acceptLargeImages: Bool
    public var primaryColor: String?
    public var hasAdminPrivileges: Bool {
        return !roles.isEmpty
    }
    private var rolesCancellable: AnyCancellable?
    private var roles: [AdminRole] = []

    /// The delegate that receives callbacks for outgoing chat events that should trigger
    /// a push notification. Held weakly — the caller is responsible for retaining it.
    public weak var pushNotificationDelegate: DittoChatPushNotificationDelegate?

    /// The hex-encoded APNs device token registered via `registerDeviceToken(_:)`.
    /// `nil` until the host app calls `registerDeviceToken(_:)`.
    public private(set) var deviceToken: String?

    private var localStore: LocalDataInterface
     var p2pStore: DittoDataInterface

    /// Posts local `UNUserNotification`s when new messages arrive in synced rooms.
    private var notificationManager: ChatNotificationManager?

    /// Retains Combine subscriptions used to keep `notificationManager` in sync with rooms.
    private var cancellables = Set<AnyCancellable>()

    init(
        ditto: Ditto?,
        retentionPolicy: ChatRetentionPolicy,
        usersCollection: String,
        userId: String?,
        userEmail: String?,
        acceptLargeImages: Bool,
        primaryColor: String?,
        pushNotificationDelegate: DittoChatPushNotificationDelegate?
    ) {
        let localStore: LocalService = LocalService()
        self.acceptLargeImages = acceptLargeImages
        self.primaryColor = primaryColor
        self.localStore = localStore
        self.p2pStore = DittoService(privateStore: localStore, ditto: ditto, usersCollection: usersCollection, chatRetentionPolicy: retentionPolicy)
        self.publicRoomsPublisher = p2pStore.publicRoomsPublisher.eraseToAnyPublisher()
        self.retentionPolicy = retentionPolicy
        self.pushNotificationDelegate = pushNotificationDelegate

        // Set up local notification manager. Subscribes to each synced room's message
        // collection via raw DittoStoreObserver so callbacks fire even when backgrounded.
        let manager = ChatNotificationManager(ditto: ditto, localStore: localStore)
        self.notificationManager = manager

        // Whenever the public rooms list changes, reconcile which rooms are being observed.
        // Use Task { @MainActor in } rather than .receive(on: DispatchQueue.main) so that
        // syncRooms — an @MainActor method — is entered through Swift's structured concurrency
        // executor. In Xcode 26.4 / Swift 6.3, calling @MainActor methods via raw GCD from a
        // nonisolated Combine sink triggers a runtime actor-isolation crash even when on the
        // main thread.
        p2pStore.publicRoomsPublisher
            .sink { [weak manager] rooms in
                Task { @MainActor [weak manager] in
                    manager?.syncRooms(rooms)
                }
            }
            .store(in: &cancellables)

        if let userId = userId {
            self.setCurrentUser(withConfig: UserConfig(id: userId))
        }
        if let email = userEmail {
            do {
                try setupRolesSubscription(email: email)
            } catch {
                // TODO: Handle errors
            }
        }
    }

    // MARK: - Push Notifications

    /// Registers an APNs device token with the SDK.
    ///
    /// Call this from `application(_:didRegisterForRemoteNotificationsWithDeviceToken:)`.
    /// The token is converted to a hex string and stored in `deviceToken` so it is
    /// accessible when building push payloads in your `DittoChatPushNotificationDelegate`
    /// implementation.
    ///
    /// - Parameter deviceToken: The raw token data provided by APNs.
    public func registerDeviceToken(_ deviceToken: Data) {
        self.deviceToken = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
    }

    /// Interprets an incoming push notification payload and returns the navigation action.
    ///
    /// Call this from your `UNUserNotificationCenterDelegate` implementation when a
    /// DittoChat-related notification arrives. The payload must contain at least
    /// `DittoChatNotificationKey.roomId` for a non-`.none` action to be returned.
    ///
    /// ```swift
    /// func userNotificationCenter(_ center: UNUserNotificationCenter,
    ///                             didReceive response: UNNotificationResponse) async {
    ///     let action = dittoChat.handleNotification(userInfo: response.notification.request.content.userInfo)
    ///     switch action {
    ///     case .openRoom(let id):      navigate(toRoomId: id)
    ///     case .openMessage(let rId, let mId): navigate(toMessageId: mId, inRoomId: rId)
    ///     case .none: break
    ///     }
    /// }
    /// ```
    ///
    /// - Parameter userInfo: The `userInfo` dictionary from the incoming notification.
    /// - Returns: A `DittoChatNotificationAction` describing what the app should do.
    public func handleNotification(userInfo: [AnyHashable: Any]) -> DittoChatNotificationAction {
        guard let roomId = userInfo[DittoChatNotificationKey.roomId] as? String else {
            return .none
        }
        if let messageId = userInfo[DittoChatNotificationKey.messageId] as? String {
            return .openMessage(roomId: roomId, messageId: messageId)
        }
        return .openRoom(id: roomId)
    }

    public var currentUserId: String? {
        get { localStore.currentUserId }
        set { localStore.currentUserId = newValue }
    }

    public func readRoomById(id: String) async throws -> Room {
        guard let room = await self.findPublicRoomById(id: id) else {
            throw AppError.unknown("room not found")
        }

        return room
    }

    func setupRolesSubscription(email: String) throws {
        try p2pStore.ditto?.sync.registerSubscription(
            query: "SELECT * FROM `roles` WHERE email = :email",
            arguments: ["email": email]
        )

        rolesCancellable = p2pStore.ditto?.store
            .observePublisher(
                query: "SELECT * FROM `roles` WHERE email = :email",
                arguments: ["email": email],
                mapTo: AdminRole.self
            )
            .catch { error in
                assertionFailure("ERROR with \(#function)" + error.localizedDescription)
                return Empty<[AdminRole], Never>()
            }
            .assign(to: \.roles, on: self)
    }

    public func allUsersPublisher() -> AnyPublisher<[ChatUser], Never> {
        p2pStore.allUsersPublisher()
    }

    public func createRoom(withConfig config: RoomConfig) async throws -> String {
        guard let id = await self.createRoom(id: config.id, name: config.name, isGenerated: config.isGenerated) else {
            throw AppError.unknown("room not found")
        }

        return id
    }

    public func createMessage(withConfig config: MessageConfig) async throws {
        let room = try await readRoomById(id: config.roomId)

        self.createMessage(for: room, text: config.message)
    }

    public func setCurrentUser(withConfig config: UserConfig) {
        self.setCurrentUser(id: config.id)
    }

    // MARK: Read
    public func read(messagesForRoom room: Room) throws {
        self.readMessagesForRoom(room: room)
    }

    public func read(messagesForUser user: ChatUser) throws {
        self.readMessagesForUser(user: user)
    }

    public func updateRoom(room: Room) throws {
        self.updateRoom(room)
    }

    /// Clears references to Ditto and running subscritopns as well as observers.
    /// Note: Make sure that you call stop sync before calling this logout function.
    public func logout() {
        notificationManager?.stopAll()
        notificationManager = nil
        cancellables.removeAll()
        p2pStore.logout()
    }
}

public class DittoChatBuilder {
    private var ditto: Ditto?
    private var retentionPolicy: ChatRetentionPolicy = .init(days: 30)
    private var usersCollection: String = "users"
    private var userId: String?
    private var userEmail: String?
    private var acceptLargeImages: Bool = true
    private var primaryColor: String?
    private var pushNotificationDelegate: DittoChatPushNotificationDelegate?

    public init() {}

    @discardableResult
    public func setDitto(_ ditto: Ditto) -> DittoChatBuilder {
        self.ditto = ditto
        return self
    }

    @discardableResult
    public func setRetentionPolicy(_ policy: ChatRetentionPolicy) -> DittoChatBuilder {
        self.retentionPolicy = policy
        return self
    }

    @discardableResult
    public func setRetentionDays(_ days: Int) -> DittoChatBuilder {
        self.retentionPolicy = ChatRetentionPolicy(days: days)
        return self
    }

    @discardableResult
    public func setUsersCollection(_ collection: String) -> DittoChatBuilder {
        self.usersCollection = collection
        return self
    }

    @discardableResult
    public func setUserId(_ id: String?) -> DittoChatBuilder {
        self.userId = id
        return self
    }

    @discardableResult
    public func setUserEmail(_ email: String?) -> DittoChatBuilder {
        self.userEmail = email
        return self
    }

    @discardableResult
    public func setAcceptLargeImages(_ accept: Bool) -> DittoChatBuilder {
        self.acceptLargeImages = accept
        return self
    }

    @discardableResult
    public func setPrimaryColor(_ color: String?) -> DittoChatBuilder {
        self.primaryColor = color
        return self
    }

    /// Sets the delegate that receives callbacks for outgoing chat events that should
    /// trigger a push notification. The delegate is held weakly by `DittoChat`.
    ///
    /// - Parameter delegate: An object conforming to `DittoChatPushNotificationDelegate`.
    @discardableResult
    public func setPushNotificationDelegate(_ delegate: DittoChatPushNotificationDelegate?) -> DittoChatBuilder {
        self.pushNotificationDelegate = delegate
        return self
    }

    @MainActor public func build() throws -> DittoChat {
        guard let ditto = ditto else {
            throw BuilderError.missingRequiredField("ditto")
        }

        return DittoChat(
            ditto: ditto,
            retentionPolicy: retentionPolicy,
            usersCollection: usersCollection,
            userId: userId,
            userEmail: userEmail,
            acceptLargeImages: acceptLargeImages,
            primaryColor: primaryColor,
            pushNotificationDelegate: pushNotificationDelegate
        )
    }

    public enum BuilderError: Error {
        case missingRequiredField(String)

        var localizedDescription: String {
            switch self {
            case .missingRequiredField(let field):
                return "Missing required field: \(field)"
            }
        }
    }
}

extension DittoChat {
    public static func builder() -> DittoChatBuilder {
        return DittoChatBuilder()
    }
}

extension DittoChat {

     func room(for room: Room) async -> Room? {
        await p2pStore.room(for: room)
    }

     func findPublicRoomById(id: String) async -> Room? {
        await p2pStore.findPublicRoomById(id: id)
    }

    public func createRoom(id: String? = UUID().uuidString, name: String, isGenerated: Bool = false) async -> String? {
        return await p2pStore.createRoom(id: id, name: name, isGenerated: isGenerated)
    }

    public func archiveRoom(_ room: Room) {
        p2pStore.archiveRoom(room)
    }

     func unarchiveRoom(_ room: Room) {
        p2pStore.unarchiveRoom(room)
    }

     func archivedPublicRoomsPublisher() -> AnyPublisher<[Room], Never> {
        localStore.archivedPublicRoomsPublisher
    }

     func readMessagesForRoom(room: Room) {
        guard let userId = currentUserId else { return }
        Task { @MainActor in
            await p2pStore.markRoomAsRead(roomId: room.id, userId: userId)
        }
    }

     func readMessagesForUser(user: ChatUser) {
        // TODO: Implement
    }
}

@MainActor
extension DittoChat {

    public func createMessage(for room: Room, text: String) {
        p2pStore.createMessage(for: room, text: text)
        pushNotificationDelegate?.dittoChat(self, didSendMessage: text, inRoom: room)
    }

    public func createImageMessage(for room: Room, image: UIImage, text: String?) async throws {
        try await p2pStore.createImageMessage(for: room, image: image, text: text)
        pushNotificationDelegate?.dittoChat(self, didSendImageMessageInRoom: room)
    }

    public func saveEditedTextMessage(_ message: Message, in room: Room) {
        p2pStore.saveEditedTextMessage(message, in: room)
    }

    public func saveDeletedImageMessage(_ message: Message, in room: Room) {
        p2pStore.saveDeletedImageMessage(message, in: room)
    }

    public func messagePublisher(for msgId: String, in collectionId: String) -> AnyPublisher<Message, Never> {
        p2pStore.messagePublisher(for: msgId, in: collectionId)
    }

    public func messagesPublisher(for room: Room, retentionDays: Int?) -> AnyPublisher<[Message], Never> {
        p2pStore.messagesPublisher(for: room, retentionDays: retentionDays)
    }

    func attachmentPublisher(
        for token: DittoAttachmentToken,
        in collectionId: String
    ) -> DittoSwift.DittoStore.FetchAttachmentPublisher? {
        p2pStore.attachmentPublisher(for: token, in: collectionId)
    }

    func createUpdateMessage(document: [String: Any?]) {
        p2pStore.createUpdateMessage(document: document)
    }

    @discardableResult
    public func fetchAttachment(
        token: [String : Any],
        deliverOn queue: DispatchQueue = .main,
        onFetchEvent: @escaping (DittoAttachmentFetchEvent) -> Void
    ) throws -> DittoAttachmentFetcher? {
        try p2pStore.ditto?.store.fetchAttachment(token: token, deliverOn: queue, onFetchEvent: onFetchEvent)
    }
}

extension DittoChat {

    var currentUserIdPublisher: AnyPublisher<String?, Never> {
        localStore.currentUserIdPublisher
    }

    public func currentUserPublisher() -> AnyPublisher<ChatUser?, Never> {
        p2pStore.currentUserPublisher()
    }

     func addUser(_ usr: ChatUser) {
        p2pStore.addUser(usr)
    }

    public func updateUser(withId id: String, name: String? = nil, firstName: String? = nil, lastName: String? = nil, subscriptions: [String: Date?]?, mentions: [String: [String]]?) {
        if let firstName, let lastName {
            p2pStore.updateUser(withId: id, name: firstName + " " + lastName, subscriptions: subscriptions, mentions: mentions)
        } else {
            p2pStore.updateUser(withId: id, name: name, subscriptions: subscriptions, mentions: mentions)
        }
    }

     func updateRoom(_ room: Room) {
        // TODO: Implement
    }

     func saveCurrentUser(name: String) {
        if currentUserId == nil {
            let userId = UUID().uuidString
            currentUserId = userId
        }

        assert(currentUserId != nil, "Error: expected currentUserId to not be NIL")

        let user = ChatUser(id: currentUserId!, name: name, subscriptions: [:], mentions: [:])
        p2pStore.addUser(user)
    }

    @available(*, deprecated, renamed: "saveCurrentUser(name:)", message: "First and last name are no needed. Use name instead")
     func saveCurrentUser(firstName: String, lastName: String) {
        if currentUserId == nil {
            let userId = UUID().uuidString
            currentUserId = userId
        }

        assert(currentUserId != nil, "Error: expected currentUserId to not be NIL")

        let user = ChatUser(id: currentUserId!, name: firstName + " " + lastName, subscriptions: [:], mentions: [:])
        p2pStore.addUser(user)
    }

     func setCurrentUser(id: String) {
        currentUserId = id
    }
}

// MARK: - Read Receipts & Unread Counts

@MainActor
extension DittoChat {

    /// Emits the count of unread messages in `room` for the current user.
    ///
    /// Derived from the current user's `subscriptions[room.id]` last-read timestamp.
    /// Messages authored by the current user and archived (tombstoned) messages are
    /// excluded. If the user has no last-read timestamp for the room (never opened it
    /// or not subscribed), the count is `0`.
    public func unreadMessagesCountPublisher(for room: Room) -> AnyPublisher<Int, Never> {
        Publishers.CombineLatest(
            messagesPublisher(for: room, retentionDays: nil),
            currentUserPublisher()
        )
        .map { messages, currentUser -> Int in
            guard let user = currentUser,
                  let lastRead = user.subscriptions[room.id] ?? nil else {
                return 0
            }
            return messages.filter { msg in
                !msg.isArchived
                    && msg.userId != user.id
                    && msg.createdOn > lastRead
            }.count
        }
        .removeDuplicates()
        .eraseToAnyPublisher()
    }

    /// Emits a `[userId: lastReadDate]` map describing which users have read up to
    /// what point in `room`. A user appears in the map only if they have a non-nil
    /// last-read timestamp for the room — i.e. they have opened it at least once
    /// since the SDK started tracking reads.
    ///
    /// To render a "read by" indicator on a message, compare each user's last-read
    /// date to the message's `createdOn`: if `lastRead >= createdOn`, the user has
    /// seen the message.
    public func readReceiptsPublisher(for room: Room) -> AnyPublisher<[String: Date], Never> {
        allUsersPublisher()
            .map { users -> [String: Date] in
                var receipts: [String: Date] = [:]
                for user in users {
                    if let lastRead = user.subscriptions[room.id] ?? nil {
                        receipts[user.id] = lastRead
                    }
                }
                return receipts
            }
            .removeDuplicates()
            .eraseToAnyPublisher()
    }
}

extension DittoChat {
     var sdkVersion: String {
        p2pStore.sdkVersion
    }

     var appInfo: String {
        let name = Bundle.main.appName
        let version = Bundle.main.appVersion
        let build = Bundle.main.appBuild
        return "\(name) \(version) build \(build)"
    }

     var peerKeyString: String {
        p2pStore.peerKeyString
    }
}

File: sdks/kotlin/src/main/java/com/ditto/dittochat/DittoChat.kt
package com.ditto.dittochat

import android.content.Context
import android.os.Bundle
import com.google.gson.Gson
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.launch
import java.util.Date
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

interface DittoChat {
    suspend fun createRoom(config: RoomConfig): String
    suspend fun createMessage(config: MessageConfig)
    fun setCurrentUser(config: UserConfig)

    val publicRoomsFlow: Flow<List<Room>>
    val hasAdminPrivileges: Boolean
    val retentionPolicy: ChatRetentionPolicy
    val acceptLargeImages: Boolean
    val primaryColor: String?

    val localStore: LocalData
    val p2pStore: DittoData
    suspend fun readRoomById(id: String): Room
    fun allUsersFlow(): Flow<List<ChatUser>>

    suspend fun updateRoom(room: Room)
    fun logout()

    // -----------------------------------------------------------------------------------------
    // Read receipts & unread counts
    // -----------------------------------------------------------------------------------------

    /**
     * Marks [roomId] as read for the current user by setting their
     * `subscriptions[roomId]` to `Date()`. Subsequent calls to
     * [unreadMessagesCountFlow] for that room emit `0` until new messages arrive.
     */
    suspend fun markRoomAsRead(roomId: String)

    /**
     * Emits the count of unread messages in [room] for the current user.
     *
     * Derived from the current user's `subscriptions[room.id]` last-read timestamp.
     * Messages authored by the current user and archived messages are excluded.
     * If the user has no last-read timestamp for the room (never opened it or not
     * subscribed), the count is `0`.
     */
    fun unreadMessagesCountFlow(room: Room): Flow<Int>

    /**
     * Emits a `Map<userId, lastReadDate>` describing which users have read up to
     * what point in [room]. A user appears in the map only if they have a non-null
     * last-read timestamp for the room — i.e. they have opened it at least once
     * since the SDK started tracking reads.
     *
     * To render a "read by" indicator on a message, compare each user's last-read
     * date to the message's `createdOn`: if `lastRead >= createdOn`, the user has
     * seen the message.
     */
    fun readReceiptsFlow(room: Room): Flow<Map<String, Date>>

    // -----------------------------------------------------------------------------------------
    // Push notifications
    // -----------------------------------------------------------------------------------------

    /**
     * Optional delegate that receives callbacks for outgoing chat events so the host app
     * can forward them to a push server (FCM). Held with a strong reference — the caller
     * controls its lifetime.
     */
    var pushNotificationDelegate: DittoChatPushNotificationDelegate?

    /**
     * The FCM registration token set by [registerDeviceToken]. Null until the host app
     * provides a token.
     */
    val deviceToken: String?

    /**
     * Stores the FCM registration token so it is available when building push payloads.
     *
     * Call this from your `FirebaseMessagingService.onNewToken` override:
     * ```kotlin
     * override fun onNewToken(token: String) {
     *     dittoChat.registerDeviceToken(token)
     * }
     * ```
     */
    fun registerDeviceToken(token: String)

    /**
     * Interprets an incoming notification's extras and returns the navigation action.
     *
     * Pass the [Bundle] from `intent.extras` (in `Activity.onCreate` or `onNewIntent`):
     * ```kotlin
     * val action = dittoChat.handleNotification(intent.extras)
     * when (action) {
     *     is DittoChatNotificationAction.OpenRoom    -> navTo("chatroom/${action.roomId}")
     *     is DittoChatNotificationAction.OpenMessage -> navTo("chatroom/${action.roomId}")
     *     is DittoChatNotificationAction.None        -> { }
     * }
     * ```
     */
    fun handleNotification(extras: Bundle?): DittoChatNotificationAction
}

@Singleton
class DittoChatImpl private constructor(
    private val context: Context,
    private val ditto: live.ditto.Ditto?,
    override val retentionPolicy: ChatRetentionPolicy,
    private val usersCollection: String,
    private val userId: String?,
    private val userEmail: String?,
    override val hasAdminPrivileges: Boolean,
    override val acceptLargeImages: Boolean,
    override val primaryColor: String?,
    override val localStore: LocalData,
    override val p2pStore: DittoData
) : DittoChat {

    override val publicRoomsFlow: Flow<List<Room>>
        get() = p2pStore.publicRoomsFlow

    var currentUserId: String?
        get() = localStore.currentUserId
        set(value) {
            localStore.currentUserId = value
        }

    // -----------------------------------------------------------------------------------------
    // Push notifications
    // -----------------------------------------------------------------------------------------

    override var pushNotificationDelegate: DittoChatPushNotificationDelegate? = null
    override var deviceToken: String? = null
        private set

    override fun registerDeviceToken(token: String) {
        deviceToken = token
    }

    override fun handleNotification(extras: Bundle?): DittoChatNotificationAction {
        val roomId = extras?.getString(DittoChatNotificationKey.ROOM_ID)
            ?: return DittoChatNotificationAction.None
        val messageId = extras.getString(DittoChatNotificationKey.MESSAGE_ID)
        return if (messageId != null) {
            DittoChatNotificationAction.OpenMessage(roomId, messageId)
        } else {
            DittoChatNotificationAction.OpenRoom(roomId)
        }
    }

    // -----------------------------------------------------------------------------------------
    // Notification manager
    // -----------------------------------------------------------------------------------------

    private val notificationManager = ChatNotificationManager(context, ditto!!, localStore)

    /** Scope used solely to collect [publicRoomsFlow] and keep [notificationManager] in sync. */
    private val notificationScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    // -----------------------------------------------------------------------------------------
    // Init
    // -----------------------------------------------------------------------------------------

    init {
        userId?.let { setCurrentUser(UserConfig(it)) }
        userEmail?.let {
            // Setup roles subscription if needed
        }

        // Keep ChatNotificationManager in sync with the live rooms list.
        notificationScope.launch {
            p2pStore.publicRoomsFlow.collect { rooms ->
                notificationManager.syncRooms(rooms)
            }
        }
    }

    // -----------------------------------------------------------------------------------------
    // DittoChat interface
    // -----------------------------------------------------------------------------------------

    override suspend fun createRoom(config: RoomConfig): String {
        val room = p2pStore.createRoom(config.id, config.name, config.isGenerated)
            ?: throw Exception("Room creation failed")
        return room
    }

    override suspend fun createMessage(config: MessageConfig) {
        val room = readRoomById(config.roomId)
        p2pStore.createMessage(room, config.message)
        pushNotificationDelegate?.dittoChat(this, config.message, room)
    }

    override fun setCurrentUser(config: UserConfig) {
        currentUserId = config.id
    }

    override suspend fun readRoomById(id: String): Room {
        return p2pStore.findPublicRoomById(id)
            ?: throw Exception("Room not found")
    }

    override fun allUsersFlow(): Flow<List<ChatUser>> {
        return p2pStore.allUsersFlow()
    }

    override suspend fun updateRoom(room: Room) {
        // Implementation needed
    }

    override fun logout() {
        notificationManager.stopAll()
        notificationScope.cancel()
        p2pStore.logout()
    }

    // -----------------------------------------------------------------------------------------
    // Read receipts & unread counts
    // -----------------------------------------------------------------------------------------

    override suspend fun markRoomAsRead(roomId: String) {
        val userId = currentUserId ?: return
        try {
            val user = p2pStore.findUserById(userId, usersCollection)
            val subs = user.subscriptions?.toMutableMap() ?: mutableMapOf()
            subs[roomId] = Date()
            p2pStore.updateUser(id = userId, subscriptions = subs)
        } catch (e: Exception) {
            // User not found yet — nothing to mark as read.
        }
    }

    override fun unreadMessagesCountFlow(room: Room): Flow<Int> {
        return combine(
            p2pStore.messagesFlow(room, null),
            p2pStore.currentUserFlow()
        ) { messages, currentUser ->
            val user = currentUser ?: return@combine 0
            val lastRead = user.subscriptions?.get(room.id) ?: return@combine 0
            messages.count { msg ->
                !msg.isArchived
                    && msg.userId != user.id
                    && (DateUtils.fromISOString(msg.createdOn)?.after(lastRead) == true)
            }
        }.distinctUntilChanged()
    }

    override fun readReceiptsFlow(room: Room): Flow<Map<String, Date>> {
        return p2pStore.allUsersFlow().map { users ->
            users.mapNotNull { user ->
                user.subscriptions?.get(room.id)?.let { date -> user.id to date }
            }.toMap()
        }.distinctUntilChanged()
    }

    fun archiveRoom(room: Room) {
        p2pStore.archiveRoom(room)
    }

    fun unarchiveRoom(room: Room) {
        p2pStore.unarchiveRoom(room)
    }

    fun archivedPublicRoomsFlow(): Flow<List<Room>> {
        return localStore.archivedPublicRoomsFlow
    }

    suspend fun createImageMessage(room: Room, imageData: ByteArray, text: String?) {
        p2pStore.createImageMessage(room, imageData, text)
        pushNotificationDelegate?.dittoChat(this, room)
    }

    suspend fun saveEditedTextMessage(message: Message, room: Room) {
        p2pStore.saveEditedTextMessage(message, room)
    }

    suspend fun saveDeletedImageMessage(message: Message, room: Room) {
        p2pStore.saveDeletedImageMessage(message, room)
    }

    fun messageFlow(msgId: String, collectionId: String): Flow<Message?> {
        return p2pStore.messageFlow(msgId, collectionId)
    }

    fun messagesFlow(room: Room, retentionDays: Int?): Flow<List<Message>> {
        return p2pStore.messagesFlow(room, retentionDays)
    }

    fun currentUserFlow(): Flow<ChatUser?> {
        return p2pStore.currentUserFlow()
    }

    suspend fun addUser(user: ChatUser) {
        p2pStore.addUser(user)
    }

    suspend fun updateUser(
        id: String,
        name: String? = null,
        subscriptions: Map<String, Date?>? = null,
        mentions: Map<String, List<String>>? = null
    ) {
        p2pStore.updateUser(id, name, subscriptions, mentions)
    }

    suspend fun saveCurrentUser(name: String) {
        if (currentUserId == null) {
            currentUserId = UUID.randomUUID().toString()
        }

        val user = ChatUser(
            id = currentUserId!!,
            name = name,
            subscriptions = emptyMap(),
            mentions = emptyMap()
        )
        p2pStore.addUser(user)
    }

    fun setCurrentUser(id: String) {
        currentUserId = id
    }

    val sdkVersion: String
        get() = p2pStore.sdkVersion

    val peerKeyString: String
        get() = p2pStore.peerKeyString

    // -----------------------------------------------------------------------------------------
    // Builder
    // -----------------------------------------------------------------------------------------

    /**
     * Fluent builder for [DittoChatImpl].
     *
     * When using Hilt, inject [Builder] directly — [Context] is provided automatically.
     * Without Hilt, use the [companion factory][DittoChatImpl.builder]:
     * ```kotlin
     * val chat = DittoChatImpl.builder(localStore, applicationContext)
     *     .setDitto(ditto)
     *     .setUserId(userId)
     *     .build()
     * ```
     */
    class Builder @Inject constructor(
        private val localStore: LocalData,
        @ApplicationContext private val context: Context
    ) {
        private var ditto: live.ditto.Ditto? = null
        var retentionPolicy: ChatRetentionPolicy = ChatRetentionPolicy(days = 30)
            private set
        var usersCollection: String = "users"
            private set
        var userId: String? = null
            private set
        var hasAdminPrivileges: Boolean = false
            private set
        var userEmail: String? = null
            private set
        var acceptLargeImages: Boolean = true
            private set
        var primaryColor: String? = null
            private set
        private var pushNotificationDelegate: DittoChatPushNotificationDelegate? = null

        fun setDitto(ditto: live.ditto.Ditto) = apply {
            this.ditto = ditto
        }

        fun setRetentionPolicy(policy: ChatRetentionPolicy) = apply {
            this.retentionPolicy = policy
        }

        fun setRetentionDays(days: Int) = apply {
            this.retentionPolicy = ChatRetentionPolicy(days = days)
        }

        fun setUsersCollection(collection: String) = apply {
            this.usersCollection = collection
        }

        fun setUserId(id: String?) = apply {
            this.userId = id
        }

        fun setUserEmail(email: String?) = apply {
            this.userEmail = email
        }

        fun setHasAdminPrivileges(hasAdminPrivileges: Boolean) = apply {
            this.hasAdminPrivileges = hasAdminPrivileges
        }

        fun setAcceptLargeImages(accept: Boolean) = apply {
            this.acceptLargeImages = accept
        }

        fun setPrimaryColor(color: String?) = apply {
            this.primaryColor = color
        }

        /**
         * Sets the delegate that receives callbacks for outgoing chat events so the host app
         * can forward them to a push server (FCM).
         */
        fun setPushNotificationDelegate(delegate: DittoChatPushNotificationDelegate?) = apply {
            this.pushNotificationDelegate = delegate
        }

        fun build(): DittoChatImpl {
            requireNotNull(ditto) { "Ditto instance is required" }

            val dittoStore = DittoDataImpl(
                privateStore = localStore,
                ditto = ditto!!,
                usersCollection = usersCollection,
                chatRetentionPolicy = retentionPolicy
            )

            return DittoChatImpl(
                context = context,
                ditto = ditto,
                retentionPolicy = retentionPolicy,
                usersCollection = usersCollection,
                userId = userId,
                userEmail = userEmail,
                hasAdminPrivileges = hasAdminPrivileges,
                acceptLargeImages = acceptLargeImages,
                primaryColor = primaryColor,
                localStore = localStore,
                p2pStore = dittoStore
            ).also { impl ->
                impl.pushNotificationDelegate = pushNotificationDelegate
            }
        }
    }

    companion object {
        /** Non-Hilt factory. [context] should be the Application context. */
        fun builder(localStore: LocalData, context: Context): Builder {
            return Builder(localStore, context)
        }
    }
}
